infra/base-images/base-builder/cargo

```diff
@@ -0,0 +1,51 @@
+#!/bin/bash -eu
+# Copyright 2020 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# This is a wrapper around calling cargo
+# This just expands RUSTFLAGS in case of a coverage build
+# We need this until https://github.com/rust-lang/cargo/issues/5450 is merged
+# because cargo uses relative paths for the current crate
+# and absolute paths for its dependencies
+#
+################################################################################
+
+export PATH="/rust/bin:$PATH"
+
+if [ "$SANITIZER" = "coverage" ] && [ $1 = "build" ]
+then
+    crate_src_abspath=`cargo metadata --no-deps --format-version 1 | jq -r '.workspace_root'`
+    export RUSTFLAGS="$RUSTFLAGS --remap-path-prefix src=$crate_src_abspath/src"
+fi
+
+if [ "$SANITIZER" = "coverage" ] && [ $1 = "fuzz" ]
+then
+    fuzz_src_abspath=`pwd`
+    export RUSTFLAGS="$RUSTFLAGS --remap-path-prefix fuzz_targets=$fuzz_src_abspath/fuzz_targets"
+    # hack to turn cargo fuzz build into cargo build so as to get coverage
+    # cargo fuzz adds "--target" "x86_64-unknown-linux-gnu"
+    (
+    # go into fuzz directory if not already the case
+    cd fuzz || true
+    # do not optimize with --release, leading to Malformed instrumentation profile data
+    cargo build --bins
+    # copies the build output in the expected target directory
+    cd target
```

Comment:
I think that this is causing a [build failure](https://oss-fuzz-build-logs.storage.googleapis.com/log-874d51ec-521f-4b99-80a6-bf8d01c6b968.txt) for wasmtime because we include the `fuzz` directory as part of the main workspace meaning that the `target` directory doesn't show up under the `fuzz` directory.

Perhaps this could use something like:

```
cargo metadata --format-version 1 --no-deps | jq '.target_directory'
```

to extract the target directory for the fuzz project?

Reply:
Another alternative, if the goal is simply to not optimize, is to set `CARGO_PROFILE_RELEASE_OPT_LEVEL=0` and that should disable all optimizations without having to override Cargo?

Reply:
Thanks @alexcrichton for debugging this cf #5366 

> Another alternative, if the goal is simply to not optimize

The goal is more than to not optimize
Indeed, we cannot optimize until https://github.com/rust-lang/rust/issues/82144 is fixed
But, if I remember correctly, `-Zinstrument-coverage` seems incompatible with other cargo options set by `cargo fuzz` (like the llvm sancov pass, which would make sense)
So we turn `cargo fuzz build` into `cargo build` to be transparent to the existing build scripts

Another solution could be to have `cargo fuzz` accept a so-called coverage sanitizer which would set `-Zinstrument-coverage`
What do you think ?

Reply:
I think ideally yeah it'd be best to codify this in `cargo fuzz` itself which could automatically apply the defaults of not optimizing and anything else necessary (like removing those extra passes)